OrcScala/src/orc/API.scala

```diff
@@ -57,80 +64,318 @@ trait OrcRuntimeProvides {
+    */
+  def halt(): Unit
+}
 
-  def schedule(t: Schedulable): Unit
+/** The state of a Future.
+  */
+sealed abstract class FutureState
 
-  def stage(ts: List[Schedulable]): Unit
+/** The future is currently not resolved.
+  */
+final case object FutureUnbound extends FutureState
 
-  // Schedule function is overloaded for convenience
-  def stage(t: Schedulable) { stage(List(t)) }
-  def stage(t: Schedulable, u: Schedulable) { stage(List(t, u)) }
+/** The future is resolved to stop.
+  */
+final case object FutureStopped extends FutureState
 
-  def stopScheduler(): Unit
+/** The future is bound to a specific value.
+  */
+final case class FutureBound(value: AnyRef) extends FutureState
 
-  def removeRoot(exec: ExecutionRoot): Unit
+/** An interface for futures or future wrappers from other systems.
+  *
+  * This interface is directly implemented by Orc futures and should be
+  * easy to implement using a wrapper on most future types. However, due
+  * to the async callback based nature of this interface futures that only
+  * provide polling or blocking interfaces will need a background thread or
+  * other machinery to implement this interface. This is unavoidable.
+  */
+trait Future {
+  /** Return the state of the future.
+    */
+  def get(): FutureState
+
+  /** Register to get the value of this future.
+    *
+    * reader methods may be called during the execution of this call (in this thread) or
+    * called in another thread at any point after this call begins.
+    */
+  def read(reader: FutureReadHandle): Unit
+}
+
+object StoppedFuture extends Future {
+  def get() = FutureStopped
+  def read(reader: FutureReadHandle) = {
+    reader.halt()
+  }
+}
+
+case class BoundFuture(v: AnyRef) extends Future {
+  def get() = FutureBound(v)
+  def read(reader: FutureReadHandle) = {
+    reader.publish(v)
+  }
+}
+
+/** Define invocation behaviors for a runtime
+  */
+trait InvocationBehavior {
+  /** Get an invoker for a specific target type and argment types.
+    *
+    * This method is slow and the results should be cached if possible.
+    *
+    * @return An Invoker or DirectInvoker for the given values or an 
+    * 			  instance of InvokerError if there is no invoker.
+    * 
+    * @see UncallableValueInvoker
+    */
+  def getInvoker(target: AnyRef, arguments: Array[AnyRef]): Invoker
+
+  /** Get an accessor which extracts a given field value from a target.
+    *
+    * This method is slow and the results should be cached if possible.
+    *
+    * @return An Accessor for the given classes or an 
+    * 			  instance of AccessorError if there is no accessor.
+    * 
+    * @see NoSuchMemberAccessor, DoesNotHaveMembersAccessor
+    */
+  def getAccessor(target: AnyRef, field: Field): Accessor
 }
 
 /** The interface through which the environment response to site calls.
+  *
+  * Published values passed to publish and publishNonterminal may not be futures.
   */
-trait Handle {
+trait Handle extends FutureReadHandle {
```

Comment:
BTW, I think `Handle` was a poor name choice back when we started OrcScala.  How about `SiteResponseChannel`?  Well maybe not "Channel", but something like that....